### DISCUSSION TITLE
Usage of Behaviour Trees and Reinforcement Learning for behaviour modelling in mesa (Gsoc 2026)

### DISCUSSION THREAD

**Discussion by sacheinn17** [mesa/mesa#3661] [Ideas]
Hi, Myself Sachein and I have just finished my mechatronics engineering. I think behaviour trees and Reinforcement Learning are ways to model complex agent behaviour beyond simple tasks.

I am a robotics programmer and I have used behaviour trees and to primarily model the behaviour of the agents. I think behaviour trees are a good fit since it has the following advantage :-

1) Easier to understand and model for complex scenarios.
2) If we can abstract the behaviour tree, we can make it so that researchers can directly share the behaviour trees which can be plugged into the step function. Making transfer easier.
3) They make the agents logic easier to inspect and debug.
4) Currently used in many complex problems for behaviour modelling of advanced agents. (Think Halo 2 and RDR 2) and also in robotics (Nav2 from Ros2).

Proposed additions: -
1) A base class for commonly used node types (sequence, selector etc...) which can be inherited by the agents
2) A way to load a tree from a common datatype (I am thinking json or xml) 
3) A simple visualization to see the tree for understand and debug the agents behaviour.

Relevant Experience : -
As mentioned, I have worked in behaviour trees in the area of robotics, here is a github link of a project in which I used the behaviourtrees.cpp package to model the behaviour of a robot. https://github.com/sacheinn17/butler_robot. The respective behaviour tree is as follows

<img width="841" height="451" alt="Screenshot from 2026-03-31 00-00-08" src="https://github.com/user-attachments/assets/744dad95-5e47-4ab3-bf7a-023bf5ba3f7d" />

# Reinforcement Learning

Implementing Reinforcement learning (especially policy gradient algorithms) allows us to model agents which can adapt unlike rule based agents. Since the ABM simulates agent-environment interaction we should be able to define a training loop pretty quickly. 

Advantages: - 
1) Allows us to model agents which cannot be modeled by a rule based algorithm
2) Provides a proper framework to train learned agents.

What's missing in mesa today: -
1) No reward abstraction
2) No state/action specification
3) The step() function is synchronous and stateless, however we need to carry state across steps to train the rl model.

Proposed additions.
1) Define a Gym like wrapper which will allows us to define the agent properly 
something like this

`
class RLAgent(Agent):
    def __init__(self):
     ...

    def get_state(self):
     ...
    def get_reward(self):
     ...

    def step(self):
        self.state = self.get_state()
        self.action = self.model.policy.select_action(self.state)
        self.apply_action(self.action)
        self.reward = self.get_reward()

    def apply_action(self, action):
     ...
`
This design is inspired by the OpenAI Gym interface — defining a clean separation between state, policy, and environment interaction — but is implemented natively within Mesa without any Gym dependency

We also need to define the model which will trigger the training loop to train the model.

2) Implement a simple Q learning based training loop
3) Implement a policy gradient method (PPO or SAC are my personal favorites to implement!).

Relevant Experience,
1) I have worked as an ML engineer at caterpillar where I worked on Reinforcement Learning (specifically deep q learning and PPO) using PyTorch.
2) I have done my bachelors project on the implementation of a bipedal walking robot based on policy gradient algorithm (PPO and SAC). 

This is my first time contributing to an opensource project and would really love some feedback on these.
